### PR TITLE
[SYCL] Support USM memcpy/copy in graphs

### LINKED ISSUE #147: Implement using handler USM memory copies in a graph node.
Implement using handler::memcpy and handler::copy for USM data inside a command-group for both record & replay and the explicit API, backend up by testing.

### PR DESCRIPTION
**Pull request by Bensuo**
- Add support for USM memcpy and copy in graphs
- New MemoryManager method for usm copies
- New PI extension methods for enqueueing USM copies
- Re-enabled and fixed issues in E2E tests for copies
- Minor changes to queue_impl to capture copies correctly

Closes #147

### CODE REVIEW

File: sycl/include/sycl/detail/pi.h [hunk 1]
```diff
@@ -2167,6 +2167,22 @@ __SYCL_EXPORT pi_result piextCommandBufferNDRangeKernel(
     const pi_ext_sync_point *sync_point_wait_list,
     pi_ext_sync_point *sync_point);
 
+/// API to append a USM memcpy command to the command-buffer.
+/// \param command_buffer The command-buffer to append onto.
+/// \param dst_ptr is the data to be copied
```
Comment (EwanC) [inline sycl/include/sycl/detail/pi.h]: think this should be `src_ptr`, and would also swap this around with `dst_ptr` to match parameter ordering

File: sycl/include/sycl/detail/pi.h [hunk 2]
```diff
@@ -2167,6 +2167,22 @@ __SYCL_EXPORT pi_result piextCommandBufferNDRangeKernel(
...
+/// \param dst_ptr is the location the data will be copied
+/// \param size is number of bytes to copy
+/// \param num_sync_points_in_wait_list The number of sync points in the
+/// provided wait list.
+/// \param sync_point_wait_list A list of sync points that this executions must
+/// wait on.
+/// \param sync_point The sync_point associated with this kernel execution.
```
Comment (EwanC) [inline sycl/include/sycl/detail/pi.h]: reword "kernel execution"

File: sycl/include/sycl/detail/pi.h [hunk 3]
```diff
@@ -2167,6 +2167,22 @@ __SYCL_EXPORT pi_result piextCommandBufferNDRangeKernel(
...
     pi_ext_sync_point *sync_point);
 
+/// API to append a USM memcpy command to the command-buffer.
+/// \param command_buffer The command-buffer to append onto.
+/// \param dst_ptr is the data to be copied
+/// \param dst_ptr is the location the data will be copied
+/// \param size is number of bytes to copy
+/// \param num_sync_points_in_wait_list The number of sync points in the
+/// provided wait list.
+/// \param sync_point_wait_list A list of sync points that this executions must
```
Comment (EwanC) [inline sycl/include/sycl/detail/pi.h]: This should be fixed in the kernel command doxy too, but "executions" should be "command"

File: sycl/test-e2e/Graph/RecordReplay/usm_copy.cpp [hunk 4]
```diff
@@ -82,12 +84,13 @@ int main() {
       CGH.ext_oneapi_graph(GraphExec);
     });
   }
-  Queue.wait_and_throw();
 
   Queue.copy(PtrA, DataA.data(), Size);
   Queue.copy(PtrB, DataB.data(), Size);
   Queue.copy(PtrC, DataC.data(), Size);
```
Comment (EwanC) [inline sycl/test-e2e/Graph/RecordReplay/usm_copy.cpp]: should these copies depend on `Event`?

File: sycl/include/sycl/detail/pi.h (hunk 1 above)
Comment (Bensuo) [reply sycl/include/sycl/detail/pi.h]: Fixed thanks!

File: sycl/test-e2e/Graph/RecordReplay/usm_copy.cpp (hunk 4 above)
Comment (Bensuo) [reply sycl/test-e2e/Graph/RecordReplay/usm_copy.cpp]: Good call, I've added it to both tests!

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 5]
```diff
@@ -8831,6 +8831,23 @@ pi_result _pi_buffer::free() {
 
 /// command-buffer Extension
 
+/// Helper function to take a list of pi_ext_sync_points and fill the provided
+/// vector with the associated ZeEvents
+pi_result getEventsFromSyncPoints(
```
Comment (EwanC) [inline sycl/plugins/level_zero/pi_level_zero.cpp]: can we use `static` or an anonymous namespace to show that this function won't be used elsewhere.

**Review by EwanC — APPROVED** [review]
LGTM, nice addittion

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 5 above)
Comment (Bensuo) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: Added `static` as that's consistent with other functions in the file!

**Review by reble — APPROVED** [review]
LGTM, updated symbols are in #187 
couldn't suggest those changes here, because the files are not in this PR.

**Pull request merged**